rfc/rfc-0000-simplify-runnable-use.md

```diff
@@ -0,0 +1,391 @@
+spec:
+   template:
+      apiVersion: tekton.dev/v1beta1
+      kind: PipelineRun
+      metadata:
+         generateName: $(runnable.metadata.name)$-pipeline-run-
+      spec:
+         pipelineRef:
+            name: linter-pipeline
+         params:
+            - name: repository
+              value: $(runnable.spec.inputs.repository)$
+            - name: revision
+              value: $(runnable.spec.inputs.revision)$
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   outputs:
+      url: spec.params[?(@.name=="repository")].value
+      revision: spec.params[?(@.name=="revision")].value
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: carto.run/v1alpha1
+      kind: Runnable
+      metadata:
+         name: $(workload.metadata.name)$-linter
+      spec:
+         runTemplateRef:
+            name: md-linting-pipelinerun
+         inputs:
+            repository: $(workload.spec.source.git.url)$
+            revision: $(workload.spec.source.git.revision)$
+         serviceAccountName: pipeline-run-management-sa
+   urlPath: .status.outputs.url
+   revisionPath: .status.outputs.revision
+```
+
+## Proposed example
+Under this proposal, users would only specify one object with a new immutable
+flag.
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterSourceTemplate    # <== Now a source template
+metadata:
+   name: source-linter
+spec:
+   template:
+      apiVersion: tekton.dev/v1beta1
+      kind: PipelineRun
+      metadata:
+         name: $(workload.metadata.name)$-linter      # <== Refers to workload
+      spec:
+         pipelineRef:
+            name: linter-pipeline
+         params:
+            - name: repository
+              value: $(workload.spec.source.git.url)$      # <== Refers to workload
+            - name: revision
+              value: $(workload.spec.source.git.revision)$ # <== Refers to workload
+         workspaces:
+            - name: shared-workspace
+              volumeClaimTemplate:
+                 spec:
+                    accessModes:
+                       - ReadWriteOnce
+                    resources:
+                       requests:
+                          storage: 256Mi
+   urlPath: spec.params[?(@.name=="repository")].value    # <== Refers to path on the immutable object that will be created
+   revisionPath: spec.params[?(@.name=="revision")].value
+   lifecycle: immutable                                         # <== Flag
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When Cartographer processes a template with the immutable flag, rather than
+creating/updating the templated object it will create/update a Runnable and
```

Comment:
have you thought about how this would impact `workload.status.resources[*].stampedRef`? I wonder if it'd be strange for someone to see in their templates that they're stamping out `kind: PipelineRun` and then they see `Runnable` there.

I particularly really like the fact that this proposal keeps that approach of the direct children being all level triggered (via the use of Runnable) though

Reply:
I discuss the .status.resources here:
https://github.com/vmware-tanzu/cartographer/blob/rfc-simplify-runnable-usage/rfc/rfc-0000-simplify-runnable-use.md#resource-reporting-on-workload

I'll note that in Office Hours Rash advocated for not creating a Runnable, and rather moving this functionality into the workload controller's code:
https://youtube.com/clip/Ugkx0s1Qy_CEr7aG2QWaiMQK1N01PRjCXNAu

I am fully amenable to that change if there is a member of the technical oversight committee that will approve that change.